zulip_bots/zulip_bots/bots/joinme/joinme.py

```diff
@@ -0,0 +1,56 @@
+from typing import Any
+import requests
+import json
+import socket
+from six.moves.configparser import SafeConfigParser
+
+class JoinmeHandler(object):
+    def usage(self) -> str:
+        return '''
+         A bot that will post a link to joinme call. Mention the bot and it will give out instructions.
+        '''
+
+    def handle_message(self, message: Any, bot_handler: Any) -> None:
+        self.config_info = bot_handler.get_config_info('joinme')
+        if 'ZISSHUDBUNIq' not in message['content']:  # The state, a special string to identify which type of url it is.
+            payload = {'client_id': self.config_info['key'], 'scope': 'user_info scheduler start_meeting',
+                       'redirect_uri': self.config_info['callback_url'],
+                       'state': 'ZISSHUDBUNIq', 'response_type': 'code'}
+            r = requests.get('https://secure.join.me/api/public/v1/auth/oauth2', params=payload)
+            if str(r) == '<Response [200]>':
+                content = "Please click the link below to log in and authorise Joinme:\n {} \nAnd please copy and " \
+                          "send the url shown in browser after clicking accept. " \
+                          "Don't forget to @-mention me!".format(r.url)
+
+            else:
+                content = "Please check if you have entered the correct API key and callback URL " \
+                          "in joinme.conf."
+        else:
+            authorization_code = message['content']
+            authorization_code = authorization_code.split('=')
+            authorization_code = authorization_code[1].split('&')
+
+            payload = {"client_id": self.config_info['key'], "client_secret": self.config_info['secret'],
+                       "code": authorization_code[0], "redirect_uri": self.config_info['callback_url'],
+                       "grant_type": "authorization_code"}
+            req = requests.post('https://secure.join.me/api/public/v1/auth/token', data = payload)
+
+            token = json.loads(req.text)
+            if 'error' in token:
```

Comment:
Can 'error' only occur when the token has expired?

Reply:
I haven't encountered any other error yet if you follow all the steps above.

Reply:
Hmm, it might be best toi check Joinme's API documentation to see their error handling.

Reply:
https://developer.join.me/docs/read/faq According to this doc, other errors occur when you refresh the token, which is not needed in this case, or the bearer is in the wrong format. But the bearer's format is fixed in this case as the codes only sub values in the correct format written in the code so there shouldn't be any problem. Therefore I  think expired token is the only one we need to consider.